osu.Server.Queues.ScorePump/ImportHighScores.cs

```diff
@@ -23,42 +24,67 @@ namespace osu.Server.Queues.ScorePump
 
         [Option(CommandOptionType.SingleValue)]
         public long StartId { get; set; }
 
+        private const int scores_per_query = 10000;
+
+        private const int seconds_between_transactions = 2;
+
         public int OnExecute(CancellationToken cancellationToken)
         {
+            Ruleset ruleset = LegacyRulesetHelper.GetRulesetFromLegacyId(RulesetId);
+            string highScoreTable = LegacyDatabaseHelper.GetRulesetSpecifics(RulesetId).HighScoreTable;
+
             using (var dbMainQuery = Queue.GetDatabaseConnection())
             using (var db = Queue.GetDatabaseConnection())
             {
-                Ruleset ruleset = LegacyRulesetHelper.GetRulesetFromLegacyId(RulesetId);
-                string highScoreTable = LegacyDatabaseHelper.GetRulesetSpecifics(RulesetId).HighScoreTable;
-                string query = $"SELECT * FROM {highScoreTable} WHERE score_id >= @startId";
+                var transaction = db.BeginTransaction();
+
+                var insertCommand = db.CreateCommand();
 
-                Console.WriteLine($"Querying with \"{query}\"");
+                insertCommand.CommandText =
+                    // main score insert
+                    $"INSERT INTO {SoloScore.TABLE_NAME} (user_id, beatmap_id, ruleset_id, data, preserve, created_at, updated_at) "
+                    + $"VALUES (@userId, @beatmapId, {RulesetId}, @data, 0, @date, @date);"
+                    // pp insert
+                    + $"INSERT INTO {SoloScorePerformance.TABLE_NAME} (score_id, pp) VALUES (@@LAST_INSERT_ID, @pp);";
 
-                var highScores = dbMainQuery.Query<HighScore>(query, new { startId = StartId }, buffered: false);
+                var userId = insertCommand.Parameters.Add("userId", MySqlDbType.UInt32);
+                var beatmapId = insertCommand.Parameters.Add("beatmapId", MySqlDbType.UInt24);
+                var data = insertCommand.Parameters.Add("data", MySqlDbType.JSON);
+                var date = insertCommand.Parameters.Add("date", MySqlDbType.DateTime);
+                var pp = insertCommand.Parameters.Add("pp", MySqlDbType.Float);
 
-                foreach (var highScore in highScores)
+                insertCommand.Prepare();
+
+                while (!cancellationToken.IsCancellationRequested)
                 {
-                    if (cancellationToken.IsCancellationRequested)
-                        break;
+                    Console.WriteLine($"Retrieving next {scores_per_query} scores starting from {StartId + 1}");
 
-                    Console.WriteLine($"Reading score {highScore.score_id}");
+                    var highScores = dbMainQuery.Query<HighScore>($"SELECT * FROM {highScoreTable} WHERE score_id > @startId LIMIT {scores_per_query}", new { startId = StartId });
 
-                    var (accuracy, statistics) = getAccuracyAndStatistics(ruleset, highScore);
+                    if (!highScores.Any())
+                        break;
```

Comment:
Not sure if it's a concern that there's two queries going on here, one for the `.Any()` and one for the `foreach`.

Reply:
It's using buffered reads now, so it won't be a second query. Not sure if it's worth doing a `ToList` or something.